File: scripts/research/bench.mjs
import { readFileSync } from 'node:fs'
import { performance } from 'node:perf_hooks'
import process from 'node:process'

const { default: AhoCorasick } = await import('../../packages/modern-ahocorasick/dist/index.js')

const patterns = readFileSync(process.argv[2], 'utf8').trimEnd().split('\n')
const text = readFileSync(process.argv[3], 'utf8')
globalThis.gc()
const before = process.memoryUsage()
let t = performance.now()
const matcher = new AhoCorasick(patterns)
const buildMs = performance.now() - t
globalThis.gc()
const after = process.memoryUsage()
function scan() {
  let count = 0
  let checksum = 0
  for (const hit of matcher.iterate(text)) {
    count++
    checksum += hit.start + hit.end + hit.patternIndex
  }
  return { count, checksum }
}
for (let i = 0; i < 20; i++) {
  scan()
  const iterator = matcher.iterate(text)
  iterator.next()
  iterator.return?.()
}
t = performance.now()
const result = scan()
const queryMs = performance.now() - t
t = performance.now()
const iterator = matcher.iterate(text)
iterator.next()
iterator.return?.()
const firstMs = performance.now() - t
console.log(JSON.stringify({ buildMs, queryMs, firstMs, ...result, retainedHeapBytes: after.heapUsed - before.heapUsed, arrayBufferBytes: after.arrayBuffers - before.arrayBuffers }))

File: scripts/research/bench.go
package main

import (
	"encoding/json"
	ac "github.com/BobuSumisu/aho-corasick"
	"os"
	"runtime"
	"strings"
	"time"
)

func main() {
	p, _ := os.ReadFile(os.Args[1])
	patterns := strings.Split(strings.TrimSuffix(string(p), "\n"), "\n")
	text, _ := os.ReadFile(os.Args[2])
	runtime.GC()
	var before, after runtime.MemStats
	runtime.ReadMemStats(&before)
	t := time.Now()
	trie := ac.NewTrieBuilder().AddStrings(patterns).Build()
	build := float64(time.Since(t).Nanoseconds()) / 1e6
	runtime.GC()
	runtime.ReadMemStats(&after)
	scan := func() (uint64, uint64) {
		var n, s uint64
		trie.Walk(text, func(end, length, id uint32) bool {
			n++
			s += uint64(end-length+1) + uint64(end+1) + uint64(id)
			return true
		})
		return n, s
	}
	for i := 0; i < 20; i++ {
		scan()
		trie.Walk(text, func(_, _, _ uint32) bool { return false })
	}
	t = time.Now()
	n, s := scan()
	query := float64(time.Since(t).Nanoseconds()) / 1e6
	t = time.Now()
	trie.Walk(text, func(_, _, _ uint32) bool { return false })
	first := float64(time.Since(t).Nanoseconds()) / 1e6
	json.NewEncoder(os.Stdout).Encode(map[string]any{"buildMs": build, "queryMs": query, "firstMs": first, "count": n, "checksum": s, "retainedHeapBytes": int64(after.HeapAlloc) - int64(before.HeapAlloc)})
	runtime.KeepAlive(trie)
}
